docs/README_API_development_workflow.md

```diff
@@ -0,0 +1,109 @@
+# API - development workflow
+
+This document outlines a proposed general workflow for the development of
+features in the API and GeoProcessing services.
+
+As the development team scales out, it is desirable to follow (flexibly) a
+workflow that allows to optimize clarity and parallel work.
+
+## Checklist
+
+### Alignment on requirements for a component
+
+* are all the project team members aware of the user-value of the component?
+* are there any assumptions implicit in the design that need clarification or
+  adjustment (because of product needs or because of technical or other
+  limitations)?
+
+### definition of the API interface(s)
+
+* which "resources" (in a REST meaning) need to be handled?
+* which endpoints are needed?
+  * should any *existing* endpoint be changed?
+  * are *new* endpoints needed?
+  * should the new endpoints share part of the URL path with existing
+    endpoints? (this is partly related to a section below on fronting
+    implementation details with a simpler public interface)
+* at this stage we should create all the relevant DTOs for create, update and
+  get payloads (requests and responses), as well as all relevant types
+* if relevant, Swagger documentation may be added now (or after the stub
+  implementation, see below)
+  
+At this stage we should have full typing as well as scaffolded endpoints. These
+would do little by now, except validating payload shape.
+
+### API e2e tests
+
+* we should aim to add at least happy-path e2e tests as one of the first
+  implementation steps
+* this means that we should create e2e tests that mimic real requests the API
+  may receive from API clients, starting from the frontend app
+* e2e tests should use DTOs and types from the previous step
+
+At this stage, we should be able to run tests for the new component and see them
+fail.
+
+### Stub implementation
+
+* consider whether a CRUD implementation is sufficient for the task
+* if so, and if queries require JSON:API query features (filtering, sorting,
+  sparse fieldsets, includes, pagination), we should base a new service on
+  `AppBaseService`
+* if the CQRS model may be more appropriate, there's still an option to go
+  very lightweight on this by only keeping command and query DTOs fully distinct
+* if CQRS proper is appropriate, it may be the way to go
+* at this stage, we should aim for a breakdown of service functions that
```

Comment:
I have added some of this in 34d352653da35da7c095d4a7f67e5a2219f9e69b.

I am not sure about *when* this fits best and I would leave this reasonably open to "what works best" - personally I find that I need to have at least a general understanding of the architecture of a component and of data and interfaces for things to make sense in my mind, but I also find that at least a rough understanding - even if not formalized - of interfaces helps me plan the "pipework" while ironing out flaws that may emerge, with full definition of final interfaces coming next.

It may also depend on context...

Reply:
I fully agree. First shape may not be final, it "just" allows everyone involved to understand how things connect to each other. It may happen that some technical difficulties arise after it was set but this happens. 
Knowing we face some external service which is hardly known, it likely could happen.
Sometimes it happens that those services enforce a way of doing things (for example, async processing / status pooling) which may not be known upfront.

As you mentioned - `what works best` and it still depends on context ;)